clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -365,6 +366,16 @@ private void replaceUnresolvedAssignmentWithNewAssignment(
      */
     @Override
     public void transitionToFenced() {
+        if (state == MemberState.PREPARE_LEAVING || state == MemberState.LEAVING) {
+            log.debug("Member {} with epoch {} got fenced but it is already leaving the group " +
+                    "with state {}, so it won't attempt to rejoin.", memberId, memberEpoch, state);
+            return;
+        }
+        if (state == MemberState.UNSUBSCRIBED) {
+            log.debug("Member {} with epoch {} got fenced but it already left the group, so it " +
+                    "won't attempt to rejoin.", memberId, memberEpoch);
+            return;
+        }
```

Comment:
This change made me think about something. My understanding is that we transition from Leaving to Unsubscribed as soon as we send the HB. How do we handle retriable errors (e.g. timeout)? My impression is that we won't retry sending it because we are in Unsubscribed.

Reply:
You got it right, we won't handle/retry anything when sending the last HB. Just a best effort to send it and then the member transitions to UNSUBSCRIBED as soon as it is sent. This is what we want right? It's the same minimal effort the legacy coordinator does on `maybeLeaveGroup`.

Reply:
I may be missing something, but I think it is the same behaviour. On `unsubscribe`, the legacy code just invokes the same `maybeLeaveGroup` ([here](https://github.com/apache/kafka/blob/96b43bf16f5625bf89b0c105a01a9f02f683ffda/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java#L1144)) that has no retry logic at all. It does have a response [handler](https://github.com/apache/kafka/blob/96b43bf16f5625bf89b0c105a01a9f02f683ffda/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java#L1166), but that's only completing/failing a future that it is ignored on the `unsubcribe`

Reply:
You're absolutely right. This is the current behaviour in the legacy consumer. I was wondering if the current behaviour is actually correct in the legacy consumer as well. Anyway, we don't have to address this here as it is fine to replicate the current behaviour. However, we should think a little more about this case, I think.

Reply:
I see, agree that makes total sense in the close (kind of "terminal" state for the consumer), but for the unsubscribe would probably make sense to put a little more effort in making sure that the member "properly" leaves the group (sending last HB). We would need to think a bit more about the implications so I filed https://issues.apache.org/jira/browse/KAFKA-15954 to review that as a follow-up.

Reply:
Thanks.